libtenzir/src/tql2/eval_binary.cpp

```diff
@@ -191,6 +192,92 @@ struct BinOpKernel<ast::binary_op::add, duration_type, time_type> {
   }
 };
 
+template <>
+struct BinOpKernel<ast::binary_op::add, duration_type, duration_type> {
+  using result = duration;
+
+  static auto evaluate(duration l, duration r)
+    -> std::variant<result, const char*> {
+    if (auto check = checked_add(l.count(), r.count()); check.has_value()) {
+      return duration{check.value()};
+    }
+    return "Duration addition overflowed";
+  }
+};
+
+template <>
+struct BinOpKernel<ast::binary_op::sub, duration_type, duration_type> {
+  using result = duration;
+
+  static auto evaluate(duration l, duration r)
+    -> std::variant<result, const char*> {
+    if (auto check = checked_sub(l.count(), r.count()); check.has_value()) {
+      return duration{check.value()};
+    }
+    return "Duration subtraction overflowed";
+  }
+};
+
+template <>
+struct BinOpKernel<ast::binary_op::div, duration_type, duration_type> {
+  using result = double;
+
+  static auto evaluate(duration l, duration r)
+    -> std::variant<result, const char*> {
+    if (r == decltype(r){}) 
+      return "division by zero";
+    return detail::narrow_cast<result>(l.count()) / detail::narrow_cast<result>(r.count());
+  }
+};
+
+template <integral_type N>
+struct BinOpKernel<ast::binary_op::mul, duration_type, N> {
+  using result = duration;
+
+  static auto evaluate(duration l, type_to_data_t<N> r)
+    -> std::variant<result, const char*> {
+    if (auto check = checked_mul(l.count(), r); check.has_value()) {
+      return duration{check.value()};
+    }
+    return "Duration multiplication overflowed";
+  }
+};
+
+template<numeric_type N> requires std::same_as<double, type_to_data_t<N>>
+struct BinOpKernel<ast::binary_op::mul, duration_type, N> {
+  using result = duration;
+
+  static auto evaluate(duration l, double r)
+    -> std::variant<result, const char*> {
+    return duration_cast<duration>(l * r);
+  }
+};
+
+template <numeric_type N>
+struct BinOpKernel<ast::binary_op::mul, N, duration_type> {
+  using result = duration;
+
+  static auto evaluate(type_to_data_t<N> l, duration r)
+    -> std::variant<result, const char*> {
+    return BinOpKernel<ast::binary_op::mul, duration_type, N>::evaluate(r, l); // Commutative
+  }
+};
+
+template <integral_type N>
+struct BinOpKernel<ast::binary_op::div, duration_type, N> {
+  using result = duration;
+
+  static auto evaluate(duration l, type_to_data_t<N> r)
+    -> std::variant<result, const char*> {
+    if (r == decltype(r){}) 
+      return "division by zero";
+
+    // NOTE: Might lose precision as the div op returns
+    // common type (aka double) but we use long as rep.
+    return std::chrono::duration_cast<duration>(l / r); 
```

Comment:
To minimize the loss of precision here you probably want to do this instead:

```cpp
using double_duration = std::chrono::duration<double, duration::period>;
return std::chrono::duration_cast<duration>(
  std::chrono::duration_cast<double_duration>(l) / r);
```

Reply:
That is still converting the `rep` to `double`, so if the underlying value for `l` is large enough, there wouldn't be any difference. Or am I missing something?

Reply:
Ah, I thought there would be a rounding difference, but apparently you're correct on  the behavior being identical. You'd need to do this to get the rounding right:

```cpp
using double_duration = std::chrono::duration<double, duration::period>;
return std::chrono::round<duration>(
  std::chrono::duration_cast<double_duration>(l) / r);
```

But then to be honest it probably doesn't matter much, if you need sub-nanosecond precision you're going to use other, more specialized tools anyways.

Reply:
That's true. I guess it's fine as-is then.